include/fastad_bits/mat.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+#include <iostream>
+#include <vector>
+
+namespace ad {
+
+template <typename T>
+class Mat;
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const Mat<T>& mat);
+template <typename T>
+bool operator==(const Mat<T>& mat1, const Mat<T>& mat2);
+
+template <typename T>
+class Mat {
+
+public:
+
+	Mat() : data(), rows_(0), cols_(0) {}
+	Mat(const Mat& mat) : data(mat.data), rows_(mat.rows_), cols_(mat.cols_) {}
```

Comment:
you may delete this line. Compiler will generate the exact same copy constructor. In fact, now that you wrote this, compiler will implicitly delete move constructor so you'd have to implement that manually as well, which is unnecessary.